### PR TITLE
#37 exact time and better intervals

### LINKED ISSUE #37: Exact time of day and better intervals between reports being sent
Should change the implementation to send, for example, everyday at 7am.

Current implementation is set to daily, weekly, 30 days (monthly) etc is just the interval being the number of seconds in that duration. Then the initial "nextReportTime" is set to 0. Every 10 mins the scheduler checks if the current time is greater than any `nextReportTime`, if there is, sends that scheduled report and increments the `nextReportTime` by the unix timestamp in `interval`.

Need to : Change UI to select an interval: Daily, Weekly, Monthly etc and also a time of day: 7am, 8am etc. and possibly, if it's weekly: What day of the week, if its monthly, what day of the month (Or potentially also "last day of month", "first of the month" etc??)

Not sure exactly what to do as there is quite a lot there. I suppose you would treat them all seperately, and have an initial step: Daily, Weekly, Monthly, then if it's daily: Time of day, otherwise weekly: Number of weeks (to handle fortnightly), day of week, then time of day. Then for monthly, number of months, day of week, time of day?

If it was done like this, I suppose you could map this into a cron scheduler string and then just start a cron job for each schedule.. but I'm not convinced this is a good idea, i'd rather save an exact time of the next report in a persistent db and check against that and increment, I think it seems safer! But determining a way to do that with these changes is more difficult, so I dunno.

### PR DESCRIPTION
**Pull request by mark-prins**
Fixes #37 

Added some inputs to the edit schedule form; so that the day of year / month / quarter can be selected and the time to send.
The day of x input only shows when year / month / quarter selected.

Time is stored as a string ( HH:mm ) which isn't ideal, but is easy to read.

Requires the addition of a `time:string` and `day:int` field to the `Schedule` table

Slight offroad.. to update the path to use `filepath.Abs(".")` rather than `..` for some reason the double dots didn't work for me 🤷‍♂️

### CODE REVIEW

**Comment by mark-prins** [comment]
also, not sure why `go.sum` was committed.

File: frontend/src/components/Schedules/EditScheduleForm.tsx [hunk 1]
```diff
@@ -54,65 +61,96 @@ export const EditScheduleForm: FC<Props> = ({ onUpdate, schedule }) => {
...
+        {(schedule.interval || 0) > 2 ? (
+          <div className={flexWrapping}>
+            <InlineFormLabel tooltip={intl.get('report_day_description')}>{intl.get('report_day')}</InlineFormLabel>
+
+            <Input
+              css=""
+              defaultValue={1}
+              min={1}
+              name={intl.get('report_day')}
+              onChange={({ currentTarget: { value } }) => onUpdate(ScheduleKey.DAY_OF_INTERVAL, parseInt(value))}
```
Comment (joshxg) [inline frontend/src/components/Schedules/EditScheduleForm.tsx:145]: wouldn't let me build without `parseInt(value, 10)` : crying about missing radix param

**Review by joshxg — CHANGES_REQUESTED** [review]
It's great! Really makes me realize how much I was over thinking the problem - love the simplicity of the solution.

Mainly the build issue - the (potentially wrong) comment about handling end of motnhs could be done later? Could add into `dbstore.go - Init()` func:

`stmt, err := db.Prepare("CREATE TABLE IF NOT EXISTS Schedule (id TEXT PRIMARY KEY, interval INTEGER, nextReportTime INTEGER, name TEXT, description TEXT, lookback INTEGER, reportGroupID TEXT, time TEXT, day INTEGER, FOREIGN KEY(reportGroupID) REFERENCES ReportGroup(id))")`

File: backend/pkg/dbstore/dbstore.go [hunk 2]
```diff
@@ -46,11 +46,13 @@ func GetDataSource() *SQLiteDatasource {
 
 	instanceManager := datasource.NewInstanceManager(getDataSourceInstanceSettings)
 
+	homePath, _ := filepath.Abs(".")
 	sqlDatasource := &SQLiteDatasource{
 		instanceManager: instanceManager,
-		Path:            filepath.Join("..", "data", "msupply.db"),
+		Path:            filepath.Join(homePath, "data", "msupply.db"),
```
[the remark this answers is not in the document]
Comment (mark-prins) [reply backend/pkg/dbstore/dbstore.go:52]: oh. yeah - I'm running from grafana root. of course! that'd be why it's different for me..
I run `./bin/darwin-amd64/grafana-server` so I have another folder to traverse. 
This is another problem that I didn't find a solution to - I'll change to `..` and update docs 😄

File: backend/pkg/dbstore/schedule.go [hunk 3]
```diff
@@ -16,23 +18,66 @@ type Schedule struct {
...
+	case 1: // weekly
+		reportTime = reportTime.AddDate(0, 0, 7)
+	default: // 0 == daily
+		if reportTime.Unix() < now.Unix() {
+			// run tomorrow
+			reportTime = reportTime.AddDate(0, 0, 1)
+		}
+	}
+	schedule.NextReportTime = int(reportTime.Unix())
+	log.DefaultLogger.Info(fmt.Sprintf("Setting time of schedule '%s' to '%s'", schedule.Name, reportTime))
```
[the remark this answers is not in the document]
Comment (mark-prins) [reply backend/pkg/dbstore/schedule.go:78]: oh, yeah. I did change how the function worked without updating the tooltips.

File: frontend/src/components/Schedules/EditScheduleForm.tsx (hunk 1 above)
Comment (mark-prins) [reply frontend/src/components/Schedules/EditScheduleForm.tsx:145]: odd that mine didn't complain. I expected that it would 🤷‍♂️

**Comment by mark-prins** [comment]
thanks @joshxg! 
73a1266

**Review by joshxg — APPROVED** [review]
Cool!

Does that work with February, though?

**Comment by mark-prins** [comment]
> Cool!
> 
> Does that work with February, though?

Yes! ( just checked on play.golang.org )

```
then := time.Date(2020,1,26,0,0,0,0,time.UTC)
	then = then.AddDate(0,2,-then.Day())
	fmt.Printf("Hello, %s", then)
```

`Hello, 2020-02-29 00:00:00 +0000 UTC`

**Pull request merged**